### ISSUE TITLE
Require an overlay

### ISSUE THREAD

**Issue by drewcrawford** [AnarchyTools/atbuild#32]
In practice, I often end up in a situation where some overlay is required.  For example, I'm building a target for either Linux or OSX, but in practice I just type `atbuild` (instead of `atbuild --overlay linux`) and so something strange happens and we don't have a good diagnostic.

We should have some way (I'm not really sure what it is) for a task to say, hey wait a minute, you can't run this task without specifying some more things.

**Comment by drewcrawford** [comment]
Proposed solution:

A task can declare

``` clojure
: requireOverlay[ ["osx" "linux"]]
```

To indicate that `osx` or `linux` is required.  The logical extension

``` clojure
: requireOverlay[ ["osx" "linux"] ["debug" "release"]]
```

specifies that the user must choose `osx` or `linux`, as well as `debug` or `release`

**Comment by owensd** [comment]
Hmm… I get the desire, but it’s not really the overlay that’s required, it’s the setting of certain properties. I could make an osx2 profile to workaround some flags, and that would be ok. I shouldn’t then need to update the task again to say, “oh, btw, osx2 is also ok."

**Comment by drewcrawford** [comment]
> I shouldn’t then need to update the task again to say, “oh, btw, osx2 is also ok."

Why not?

**Comment by owensd** [comment]
Because overlays are just ways to specify settings; there’s nothing inherently special about the name of the overlay that happens to specify a set of settings.

What you really have is a set of properties that must be set to continue, how they get set shouldn’t be important. They may get overridden by a user’s profile, the command line, or from an imported package file’s overlays.

**Comment by drewcrawford** [comment]
> What you really have is a set of properties that must be set to continue,

False.

Let's take the example of the commit referenced in this issue.  I will develop the entire example, working my way back.
1.  A build failed
2.  The build failed because `AnarchyDispatch` could not be imported:
   
   ```
    NaOH/SwiftAPI.swift:16:12: error: no such module 'AnarchyDispatch'
   import AnarchyDispatch
          ^
   <unknown>:0: error: build had 1 command failures
   ```
3.  AnarchyDispatch could not be imported because the `compileFlags` `["-I" "AnarchyDispatch/anarchy"]` were not present
4.  These compile flags are supposed to come in by activating the `AnarchyDispatch.compile-osx` overlay (or `AnarchyDispatch.compile-linux` overlay), as specified in AnarchyDispatch [here](https://github.com/AnarchyTools/AnarchyDispatch/blob/master/build.atpkg#L4)
5.  The failing project (one level up) activates that overlay in the case that _its_ `osx` or `linux` overlay are selected [here](https://code.sealedabstract.com/drewcrawford/NaOH/blob/master/build.atpkg#L21)
6.  I built from the CLI without specifying either of those overlays.

Now returning to your statement:

> What you really have is a set of properties that must be set to continue,

The only property involved, `compileFlags` _was_ set.  [Here](https://code.sealedabstract.com/drewcrawford/NaOH/blob/master/build.atpkg#L19) is what it was set to.  And that is the _correct_ thing to set it to; it does not know anything about AnarchyDispatch or its paths, or should it.  That information comes in from AnarchyDispatch's atpkg, not from the parent project.

Further, if we _already_ have a syntax to say "NaOH must be compiled with `-I AnarchyDispatch/anarchy` it is simply

``` clojure
:tasks {
    :NaOH {
           :compileFlags ["-I" "AnarchyDispatch/anarchy"]
    }
}
```

which is a syntax that not only specifies that it should be done, but actually does it, which is superior.  But there is a reason we don't use that syntax here: specifying that the flag _should_ be present is not what we want.

The problem really is that these tasks (both NaOH, and `AnarchyDispatch.default`) are "abstract" (in the OO sense), they are not intended to be used without an appropriate overlay, that provides the "concrete" implementation for the platform.  Using them without the intended overlay is user error, and we should have a diagnostic to catch this case, as it's about the 5th time I've done it.

**Comment by owensd** [comment]
> > What you really have is a set of properties that must be set to continue,
> 
> False.

True. ;)

> Let's take the example of the commit referenced in this issue.  I will develop the entire example, working my way back.
> 1.  A build failed
> 2.  The build failed because `AnarchyDispatch` could not be imported:
>    
>    ```
>    NaOH/SwiftAPI.swift:16:12: error: no such module 'AnarchyDispatch'
>    import AnarchyDispatch
>           ^
>    <unknown>:0: error: build had 1 command failures
>    ```
> 3.  AnarchyDispatch could not be imported because the `compileFlags` `["-I" "AnarchyDispatch/anarchy"]` were not present
> 4.  These compile flags are supposed to come in by activating the `AnarchyDispatch.compile-osx` overlay (or `AnarchyDispatch.compile-linux` overlay), as specified in AnarchyDispatch [here](https://github.com/AnarchyTools/AnarchyDispatch/blob/master/build.atpkg#L4)
> 5.  The failing project (one level up) activates that overlay in the case that _its_ `osx` or `linux` overlay are selected [here](https://code.sealedabstract.com/drewcrawford/NaOH/blob/master/build.atpkg#L21)
> 6.  I built from the CLI without specifying either of those overlays.
> 
> Now returning to your statement:
> 
> > What you really have is a set of properties that must be set to continue,

Would it have been better if I had said, "a set of properties that must be set and contain the following values?"

> The only property involved, `compileFlags` _was_ set.  [Here](https://code.sealedabstract.com/drewcrawford/NaOH/blob/master/build.atpkg#L19) is what it was set to.  And that is the _correct_ thing to set it to; it does not know anything about AnarchyDispatch or its paths, or should it.  That information comes in from AnarchyDispatch's atpkg, not from the parent project.
> 
> Further, if we _already_ have a syntax to say "NaOH must be compiled with `-I AnarchyDispatch/anarchy` it is simply
> 
> ``` clojure
> :tasks {
>   :NaOH {
>          :compileFlags ["-I" "AnarchyDispatch/anarchy"]
>   }
> }
> ```
> 
> which is a syntax that not only specifies that it should be done, but actually does it, which is superior.  But there is a reason we don't use that syntax here: specifying that the flag _should_ be present is not what we want.

That's not how it works. That task sets the `compileFlags` to a value. However, that value can be overwritten by overlays. That functionality isn't implemented yet to overwrite arrays and maps, but it does exist in the `llbuild` branch for literals as you should absolutely be able to overwrite the values set in tasks.

> The problem really is that these tasks (both NaOH, and `AnarchyDispatch.default`) are "abstract" (in the OO sense), they are not intended to be used without an appropriate overlay, that provides the "concrete" implementation for the platform.  Using them without the intended overlay is user error, and we should have a diagnostic to catch this case, as it's about the 5th time I've done it.

Again, what you want is to say that my task only works if certain tasks are set or contain a set of values.

``` clojure
:validation {
  :compileFlags {:contains ["-I AnarchyDispatch/anarchy"] }
}
```

Requiring a specific overlay is no guarantee that the properties are going to have values that you need to have set. At best, it's a loose contract that says, "yeah, those overlays I named should work, but I'm going to make you guess at what I really require from them vs. what are just set as convenience items".

Think of overlays as transparent pieces of paper that stack on top of each to make a full image. Sometimes parts are overwritten, sometimes things are merged, and sometimes things are added.

Here's an example of why what you propose is not sufficient:

Let's say I'm working on a new Foundation layer. This layer is cross-platform but fairly problematic for the time being, so it goes in an `experimental` overlay. I still want to specify the `osx` and `linux` options as there are are flags I get from them that I care about. However, this `experimental` overlay needs to replace some values in the `osx` overlay and adds some new properties. If the `experimental` override replaces the `compileFlags` and doesn't add your `AnarchyDispatch/anarchy`, your check for the usage of `osx` had absolutely no effect.

**Comment by drewcrawford** [comment]
We're talking past each other.  What I said was:

> anyone who builds my app must choose an overlay

You're telling me that the requirement I actually wanted is 

> anyone who links with AnarchyDispatch must include `-I AnarchyDispatch/anarchy`

That modification is more or less correct, in some kind of perfect world.  It is also useless, for two reasons.

First, consider the dependent clause.  If it is true that anyone who links with AnarchyDispatch _must_ include the flags, then we should simply not allow these flags to **not** be included, stated alternately, we do not allow you to remove them.  (Via an overlay or any other way.)  Pick your implementation: we either lock them in, or we provide a warning, or we error to the user, but in any case, there is no way to move forward.

An outcome in which it is impossible to override the flags seems to me absurd; in fact your hypothetical assumes there is some syntax to do this.  Why you are surprised that a feature you designed to remove flags works as you designed it I do not know, unless you intend to forbid completely the removal of these flags, which I would object to.  In any case, I am not surprised or troubled by this scenario.  It's a feature.

Second, considering the predicate clause: `anyone who links with AnarchyDispatch`.  There is no general-purpose algorithm to determine if I link with AnarchyDispatch or not.  Proof: you might imagine we could parse link flags for `-L`, but there is also the different syntax `-l`, there is also `-Xlinker /path/to/foo.a` and so on, and then consider the case of a more arbitrary compiler than clang, and a more arbitrary tool than atllbuild, the result follows.

So in summary, I do _not_ want to check to see if those flags were included (because overriding them in your `experimental` layer is a feature, not a bug), and I also do _not_ want to condition this on the linking of AnarchyDispatch (because that is not computable).  I really do want the feature I originally stated: to declare that a task requires some overlay(s).  It is not a perfect model for the problem (and there isn't one), but it is a much closer fit than this line of inquiry is.

**Comment by owensd** [comment]
I agree that this solves a pragmatic problem and I'm not completely against adding this; I've thought about it before too. 

I'd make a minor change to the syntax:

``` clojure
:require-overlays [["osx" "linux"] ["debug" "release"]]
```

**Comment by drewcrawford** [comment]
I thought of another case of this from my next project lying in bed, and I wanted to write it up before it slipped out of my head.  This case has nothing to do with platforms.

Let's say we have a target/task, and its dependencies are "tedious".  Choose your own adventure--they take a long time to compile, they require Rust to be installed, whatever.  Arbitrarily tedious.

So, instead of compiling from source, you write for each dependency a task that fetches a precompiled binary with curl for example.

But then, sometimes you need to develop the dependencies (even though that is tedious, sometimes it is necessary) and so we need some way to switch between precompiled and built modes.

Overlays to the rescue!  So we have one overlay for `precompiled-dependencies` and another overlay for `build-dependencies`.  And as implementation we change the `dependencies` for our build task via the overlays (to point to to the download or build tasks respectively).  (It is also possible, and equivalent, to change the implementation of those tasks themselves, e.g. from shell/curl to shell/make for example.)

The problem is, there is no "value" for us to "validate".  In _every_ case, we want to link with the dependencies.  So of course we just specify them without an overlay in a perfectly ordinary way:

``` clojure
:build {
    :linkWithProduct ["Foo.a" "non-tedious-library.a"]
    :dependencies ["non-tedious-library"]
    :overlay {
         :precompiled-dependencies {:dependencies ["foo-precompiled"] }
         :build-dependencies {:dependencies ["foo-buildfromsource"] }
    }
}
```

To further the analogy against "validation", the thing we want to "validate" here is that `Foo.a` will appear in the filesystem by the time `build` is actually run.  But obviously we have no way to validate that; curl may fail, the dependency may not compile, etc.  It is the halting problem; we cannot determine what a task will do on the file system, even though that perhaps more closely matches the guarantee we would like.

What we can do, and what is pragmatic to do, is verify that at least one of those overlays is present if `build` is going to be built.  As long as one of those overlays is present, we will at least try to do something intelligent, and we will not get halfway into running `build` in a way that is hopeless without its dependency present.

**Comment by owensd** [comment]
Yep, sounds good. 

To be clear, I don't actually think the validate "solution" is a good one.

**Comment by owensd** [comment]
The more I think about this, the more I think we should just add `platforms` and `build-configs` (build flavors, debug, release, optimized, blah).

Then we can just have an internal rule that states that both of those must be present, if the package doesn't define the defaults.

They have the exact same syntax as `overlays`; they'd just be another property on `Package` and they would _not_ participate in merging values like `overlays` do.

**Comment by drewcrawford** [comment]
> Then we can just have an internal rule that states that both of those must be present,

That doesn't solve my second example though.  In that case we want to specify 1.  platform, 2.  debug/release (presumably) and 3.  build-dependencies or download-dependencies.

**Comment by owensd** [comment]
Sorry, my above comment did not mean to negate this feature as well, just that platforms and configs might be better handled outside.

**Comment by owensd** [comment]
I'm also not 100% sure that we should have `platforms`, but we can move this discussion outside of this thread. 

+1 for this bug as-is.
